### ISSUE TITLE
BN update

### ISSUE THREAD

**Issue by wuzhi19931128** [JaminFong/FNA#10]
I use the code to verify the fact that when the input of BN is 0,then the grad of BN is also 0 and the parameter of conv can't update.
What special treatment does your code have to make it updatable? 
Thank you very much for your answer.

[test(2).txt](https://github.com/JaminFong/FNA/files/5223822/test.2.txt)

**Comment by JaminFong** [comment]
Hi, I have run your test code and find that the parameters of both convs and bn in `mm` have been updated.
In addition, the affine parameters of bn in our implementation are not initialized with zeros.
https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/tools/apis/param_remap.py#L47-L50

**Comment by wuzhi19931128** [comment]
Sorry , the BN are initilalized withs zeros because I find the zero weights in conv did't be updated. I change it in the following code .  I compare the weighr before training and after training, the weight of original channels of conv can be updated but new 
channels can't.

import torch
import torch.nn as nn

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3,6,1,bias=False)
        self.bn1 = nn.BatchNorm2d(6)
        self.relu = nn.ReLU(inplace=True)
        self.conv2 = nn.Conv2d(6,1,1,bias=False)

    def forward(self,x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)

        x = self.conv2(x)

        return x

m = Net()
input = torch.randn(1,3,3,3)


mm = Net()
mm.conv1 = nn.Conv2d(3,9,1,bias=False)
mm.bn1 = nn.BatchNorm2d(9)
mm.conv2 = nn.Conv2d(9,3,1,bias=False)


conv1_weight = torch.zeros_like(mm.conv1.weight, requires_grad=True)#mm.conv1.weight.data#
conv2_weight = torch.zeros_like(mm.conv2.weight, requires_grad=True)#mm.conv2.weight.data#
bn1_weight = mm.bn1.weight.data
bn1_bias = mm.bn1.bias.data

conv1_weight.narrow(0, 0, 6).copy_(m.conv1.weight)
conv2_weight.narrow(0,0,1).narrow(1,0,6).copy_(m.conv2.weight)
bn1_weight.narrow(0, 0, 6).copy_(m.bn1.weight)
bn1_bias.narrow(0, 0, 6).copy_(m.bn1.bias)


states = mm.state_dict()
states["conv1.weight"] = conv1_weight
states["conv2.weight"] = conv2_weight
states["bn1.weight"] = bn1_weight
states["bn1.bias"] = bn1_bias

mm.load_state_dict(states)
print(states['conv2.weight'][0,:,0 , 0])
print(states['conv1.weight'][:, 0, 0 , 0])

o3 = mm(input)
for i in range(50):
    optimizer = torch.optim.SGD(mm.parameters(), 0.1)
    optimizer.zero_grad()
    loss = mm(input).sum()
    loss.backward()
    optimizer.step()
o4 = mm(input)

states = mm.state_dict()
print(states['conv2.weight'][0,:,0 , 0])
print(states['conv1.weight'][:, 0, 0 , 0])
print(states['bn1.weight'])
print(states['bn1.bias'])`

the output is:
`tensor([-0.1885,  0.2722, -0.3236, -0.0080,  0.1743,  0.1851,  0.0000,  0.0000,
         0.0000], grad_fn=<SelectBackward>)
tensor([ 0.3814, -0.3743, -0.5190, -0.1403,  0.1095, -0.4884,  0.0000,  0.0000,
         0.0000], grad_fn=<SelectBackward>)
tensor([-9.4903e+18,  1.9708e-01, -1.6049e+19, -7.7098e+18, -1.0708e+19,
        -1.2404e+18,  0.0000e+00,  0.0000e+00,  0.0000e+00])
tensor([ 1.2955e+12, -3.7314e-01,  2.1929e+11, -6.9362e+11,  3.5190e+10,
         3.0221e+11,  0.0000e+00,  0.0000e+00,  0.0000e+00])
tensor([ 7.1313e+10,  6.5727e-02, -1.0795e+11, -2.4408e+11,  1.3151e+11,
        -8.9295e+09,  7.8840e-01,  1.0940e-01,  4.5199e-01])
tensor([ 1.6438e+19, -1.0287e-01,  2.7798e+19,  1.3354e+19,  1.8547e+19,
         2.1484e+18,  0.0000e+00,  0.0000e+00,  0.0000e+00])
`

the weights of conv and bn indeed be updated, but the 0 in conv is still 0.  
Sorry for vague expression inthe  first question.
[test.txt](https://github.com/JaminFong/FNA/files/5229302/test.txt)

**Comment by JaminFong** [comment]
Hi, sorry for the unclearness. The `zeros_like` initialization for new parameters in our implementation is actually targeted at the kernel level which will not cause the non-update. As the search space only involves the remapping to a narrower width (i.e. smaller or equal expansion ratios compared to the seed network MobileNetV2), this operation will not cause 0-assigned new parameters on the width level. If remapping to a larger width is needed, a random initialization is recommended. Hoping this could help you.
https://github.com/JaminFong/FNA/blob/898ec5d3f24c3128cf1a4bb276a542f1fab3bad5/fna_det/tools/apis/param_remap_search.py#L46

**Comment by wuzhi19931128** [comment]
Thanks. This is a little doubt when I read the paper. Now I get it.
 I found that all parameters can be updated if we use 0 to initialize the new channels of BN  but randomly  initialize the new channels of conv. It's just a little change  but it  increases the selectable space of the model.
